db/collection.go

```diff
@@ -196,13 +211,108 @@ func (c *Collection) Find(q *Query, opts ...TxnOption) (instances [][]byte, err
 	return
 }
 
+// validInstance validates the json object against the collection schema.
+func (c *Collection) validInstance(v []byte) error {
+	r, err := gojsonschema.Validate(c.schemaLoader, gojsonschema.NewBytesLoader(v))
+	if err != nil {
+		return err
+	}
+	errs := r.Errors()
+	if len(errs) == 0 {
+		return nil
+	}
+	var msg string
+	for i, e := range errs {
+		msg += e.Field() + ": " + e.Description()
+		if i != len(errs)-1 {
+			msg += "; "
+		}
+	}
+	return fmt.Errorf("%w: %s", ErrInvalidSchemaInstance, msg)
+}
+
+// validWrite validates new events against the identity and user-defined write validator function.
+func (c *Collection) validWrite(identity thread.PubKey, e core.Event) error {
+	if c.writeValidator == nil {
+		return nil
+	}
+	vm := goja.New()
+	validate, writer, err := getJSFunc(vm, c.writeValidator, identity)
+	if err != nil {
+		return err
+	}
+	data, err := e.Marshal()
+	if err != nil {
+		return err
+	}
+	event, err := parseJSON(vm, data)
+	if err != nil {
+		return err
+	}
+	var inv goja.Value
+	val, err := c.db.datastore.Get(baseKey.ChildString(c.name).ChildString(e.InstanceID().String()))
+	if err != nil && !errors.Is(err, ds.ErrNotFound) {
+		return err
+	}
+	if val != nil {
+		inv, err = parseJSON(vm, val)
+		if err != nil {
+			return err
+		}
+	}
+	res, err := validate(nil, writer, event, inv)
+	if err != nil {
+		return fmt.Errorf("error running write validator func: %v", err)
+	}
+	out := res.Export()
+	switch out.(type) {
+	case bool:
+		if out.(bool) {
+			return nil
+		} else {
+			return app.ErrInvalidNetRecordBody
+		}
+	case nil:
+		return app.ErrInvalidNetRecordBody
+	default:
+		return fmt.Errorf("%w: %v", app.ErrInvalidNetRecordBody, out)
+	}
+}
+
+// filterRead filters an instance against the identity and user-defined read filter function.
+func (c *Collection) filterRead(identity thread.PubKey, instance []byte) ([]byte, error) {
+	if c.readFilter == nil {
+		return instance, nil
+	}
+	vm := goja.New()
+	filter, reader, err := getJSFunc(vm, c.readFilter, identity)
+	if err != nil {
+		return nil, err
+	}
+	inv, err := parseJSON(vm, instance)
+	if err != nil {
+		return nil, err
+	}
+	res, err := filter(nil, reader, inv)
+	if err != nil {
+		return nil, fmt.Errorf("error running read filter func: %v", err)
+	}
+	out := res.Export()
+	switch out.(type) {
+	case nil:
+		return nil, nil
+	default:
+		return json.Marshal(out)
```

Comment:
Should have a safer type checking as done in the other function?
(i.e: a `default` case that errors, and only unmarshaling in the correct output type)

Reply:
We could check for whether or not `out` is _not_ a bool, string, int, etc... since that would be weird to have a primitive _as_ the instance, but I that feels a little heavy handed given that it is valid JSON still. What do you think?

Reply:
Dismiss my comment... afterward, when I see that filter can mutate things, I got a better sense of what this return is. Looks good to me as is!